README.md
# README.md

## CMPE 195 Sp'21 - Fall'22 Year Project
Project by Andy Ho, Austin Nguyen, Huey Nguyen

## How to run
Downloading and running your own version of the Discord bot is unnecessary.

To install the Discord bot onto your own server, follow this link: https://discord.com/oauth2/authorize?client_id=949120643557900328&permissions=1152125172800&scope=bot


changelog.md
# Changelog
All notable changes to this project will be documented in this file using Semantic Versioning 2.0.0.

## [1.6.6] - 2022-04-13
### Added
- Quoting now also displays the day's change in value and percentage
- Quoting functionality improved to simplify function in main bot file
### Changed
- Quoting displays the same name and symbol for ETFs
### Fixed
- Fixed the stock name lookup for ETFs
### Removed
- main.py removed


## [1.6.5] - 2022-04-11
### Changed
- Look up stock quotes now uses Finnhub API function to simplify functionality
- Charting functionality improved to simplify function in main bot file
### Fixed
- Fixed 403 Error when looking up stock quotes with Finnhub API function
### Removed
- Quoting no longer needs to use requests library to fetch stock information from Finnhub

## [1.6.4] - 2022-04-10
### Changed
- import Discord library to send embedded messages from within function


## [1.6.3] - 2022-04-08
### Added
- Quoting functionality to look up daily stock information


## [1.6.2] - 2022-03-25
### Added
- import Finnhub api with token


## [1.6.1] - 2022-03-18
### Added
- Heads or Tails function for some fun lottery plays


## [1.6.0] - 2022-03-11
### Added
- Created quote.py for finnhub stock quoting
- !sheesh function for fun
- randomize emoticons within sheesh function


## [1.5.3] - 2022-03-04
### Added 
- Download charts into /charts directory
- Charting( ) also deletes file after being sent to Discord
### Changed
- Send chart method changed from posting link to sending chart.png
### Fixed
- No longer displays link to chart from Finviz when a user tries to download or sees a notification from Discord


## [1.4.0] - 2022-03-03
Meeting with Project Advisor
### Added
- Send link to chart from Finviz with specified symbol into Discord chat
### Changed
- Commands begin with '!' instead of '$'
- !help message sent to chat instead of Discord DMs


## [1.3.0] - 2022-02-25
### Added
- chart.py for chart fetching function
### Changed
- Bot will send some messages as an embedded message for readability


## [1.2.1] - 2022-02-18
### Added
- bot.py for main bot functionality
### Changed
- Get stock information from any US market stock symbol with command
### Removed
- main.py removed to make Discord bot functions modular and readability
- Specific function for GME and AMC information fetch


## [1.1.0] - 2022-01-28
### Added
- Stock and options information fetch on 'GME' with command
### Changed
- Presentation of stock and options information with tabulate for readability


## [1.0.0] - 2022-01-14
### Added
- Stock information fetch on 'AMC' with command
- Options information fetch on 'AMC' with command


## [0.4.1] - 2021-12-20
### Added
- Basic Discord bot functionality of sending messages on command
- Import libaries for stock information


## [0.3.2] - 2021-12-17
### Added
- Discord command parsing
### Changed
- Initializing Discord client with Tickr Bot token


## [0.2.3] - 2021-11-26
### Added
- Tickr Bot created on Discord
- Tickr Bot token set as secret environment variable
- Discord bot linked to test channel


## [0.1.1] - 2021-10-29
### Added
- main.py created
- import discord


main.py
# Run command in shell when starting project, python3 -m pip install -U discord.py

import discord
from discord.ext import commands
from tabulate import tabulate
import os
import yfinance as yf

# bot = discord.ext.commands.Bot(command_prefix = "!");

client = discord.Client()

amc = yf.Ticker('AMC')
price = str(amc.info['currentPrice'])
sector = amc.info['sector']

# ------ Earliest Options Expiration ---------

op = amc.options
earlyExpiration = op[0]
opt = amc.option_chain(earlyExpiration)
calls = opt.calls.head()

# calls.type
# df = calls.DataFrame({'Date': [calls['contractSymbol']],'Strike':[calls['strike']]})

gme = yf.Ticker('GME')
gprice = str(gme.info['currentPrice'])
gsector = gme.info['sector']
gsumm = gme.info['longBusinessSummary']

op1 = gme.options
# earlyExpiration = op1[0]
opt1 = gme.option_chain(earlyExpiration)
calls1 = opt1.calls.head()


@client.event
async def on_ready():
    print('We have logged in as {0.user}'.format(client))


@client.event
async def on_message(message):
    if message.author == client.user:
        return

    if message.content.startswith('!AMC'):
        await message.channel.send('Current Price of AMC: $' + price)
        await message.channel.send("Sector: " + sector)
        await message.channel.send("Call options: ")
        print(calls)
        await message.channel.send(calls)

        # await message.channel.send(tabulate(calls, headers='keys', tablefmt='psql'))

    if message.content.startswith('!GME'):
        await message.channel.send('`Current Price of GME: $' + gprice + '`')
        await message.channel.send("`Sector: " + gsector + '`')
        await message.channel.send("Call options: ")
        await message.channel.send(calls1)
        await message.channel.send("`Summary: " + gsumm + '`')


#------------------- User Command Parsing ---------------------
# @commands.command()
# async def test(ctx, arg):
#     await ctx.send(arg)

# bot.add_command(test)

# @bot.command()
# async def test(bot, arg):
#     await bot.send(arg)

client.run(os.environ['TOKEN'])
# bot.run(os.environ['TOKEN'])


chart.py
import discord
import requests
import os


def get_chart(symbol, duration):
    temp_duration = 'i5'
    duration_options = ['M','m','W','w','D','d','15','5','3']
    if(duration == 'M' or duration == 'm'):
        temp_duration = 'M'
    elif(duration == 'W' or duration == 'w'):
        temp_duration = 'W'
    elif(duration == 'D' or duration == 'd'):
        temp_duration = 'D'
    elif(duration == '15'):
        temp_duration = 'i15'
    elif(duration == '5'):
        temp_duration = 'i5'
    elif(duration == '3'):
        temp_duration = 'i3'          

    daily_indicators = 'rsi_b_14,macd_b_12_26_9,ema_9,ema_21,sma_50,sma_200,sma_20'
    indicators = daily_indicators if (temp_duration == 'D') else '0'

    headers = {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.77 Safari/537.36"}
    
    url = f'https://elite.finviz.com/chart.ashx?t={symbol}&ty=c&ta={indicators}&p={temp_duration}&s=l'

    if (duration not in duration_options):
        invalid_duration = discord.Embed(
            color = discord.Color.red(),
            title = 'Invalid time period.',
            description = '!help chart for more info.'
        )
        return invalid_duration
    else:
        response = requests.get(url, headers=headers)
        if not os.path.exists('charts'):
            os.mkdir('charts')
    
        with open(os.path.join('charts', f'chart_{symbol}{duration}.png'), 'wb') as handle:
            handle.write(response.content)
        
        return True
        

quote.py
import os
import finnhub
import discord

finnhub_client = finnhub.Client(api_key=os.environ['finnhub'])

not_found = discord.Embed(
    color = discord.Color.red(),
    title = "Stock symbol does not exist.",
    description = "!help quote for more info."
)

def get_quote(symbol):    
    stock_quote = finnhub_client.quote(symbol={symbol.upper()})
    if stock_quote['c'] == 0:
        return not_found
    else:
        if stock_quote['d'] > 0:
            change_color = discord.Color.green()
        elif stock_quote['d'] < 0:
            change_color = discord.Color.red()
        else:
            change_color = discord.Color.light_grey()
    
    stock_name = finnhub_client.company_profile2(symbol={symbol.upper()})
    if stock_name == {}:
        name = symbol.upper()
    else:
        name = stock_name['name']
    
    embed = discord.Embed(
        color = change_color,
        title = f'{name} ({symbol.upper()}) Quote' 
    )
    
    embed.add_field(name = 'Current Price:', value = f"{round(stock_quote['c'], 2)}", inline = False)
    embed.add_field(name = 'Open:', value = f"{round(stock_quote['o'], 2)}", inline = False)
    embed.add_field(name = 'Today\'s Change:', value = f"{round(stock_quote['d'], 2)} ({round(stock_quote['dp'], 2)}%)", inline = False)
    embed.add_field(name = 'Today\'s High:', value = f"{round(stock_quote['h'], 2)}", inline = False)
    embed.add_field(name = 'Today\'s Low:', value = f"{round(stock_quote['l'], 2)}", inline = False)
    
    return embed

bot.py
import os
import discord
from discord.ext import commands
import chart as c
import quote as q

import random

description = 'TickrBot - A Python Discord Bot made by Group #24 of Sp\'22'
bot = commands.Bot(command_prefix='!', description=description)
bot.remove_command('help')  # Help command


# Shows Discord bot online in console
@bot.event
async def on_ready():
	activity = discord.Game(name="!help for commands", type=3)
	await bot.change_presence(status=discord.Status.online, activity=activity)
	print(f'{bot.user.name} logged in.')


# Shows Discord bot offline in console
@bot.event
async def on_disconnect():
	print(f'{bot.user.name} disconnected')


# Some error handling
@bot.event
async def on_command_error(ctx, error):
	embed = discord.Embed(color=discord.Color.red())
	if isinstance(error, commands.MissingRequiredArgument):
		embed.add_field(name='Missing an argument!',
		                value='See !help for help.')
		await ctx.send(embed=embed)


# !help <command> function
@bot.command(name='help')
async def help_command(ctx, command=None):
	author = ctx.message.author

	if command != None:
		embed = discord.Embed(color=discord.Color.blue(),
		                      title='TickrBot Help')

		# Add !help commands here - alphabetize please
		if command == 'chart':
			embed.add_field(
			    name='!chart <symbol> <duration>',
			    value=
			    'Generate a chart for given symbol in given duration. Durations: M, W, D, 15, 5, 3 are for the monthly, weekly, daily, 15, 5, and 3 minute.'
			)

		elif command == 'help':
			embed.add_field(name='Help Command', value='Help with the help?')

		elif command == 'quote':
			embed.add_field(name='!quote <symbol>',
			                value='Get information about a specific stock.')

		else:
			embed.add_field(name='Unknown Command',
			                value='Command not recognized.')

		await ctx.send(embed=embed)

	# List basic help info
	elif command == None:
		embed = discord.Embed(
		    color=discord.Color.blue(),
		    title=
		    'TickrBot Help \n!help <command> for help on a specific command.')
		embed.add_field(name='!chart <symbol> <duration>',
		                value='Gets a stock chart in a specified duration.',
		                inline=False)
		embed.add_field(name='!help <command>',
		                value='Get help with !help <command>.',
		                inline=False)
		embed.add_field(name='!quote <symbol>',
		                value='Gets daily information on the specific symbol.',
		                inline=False)
		embed.set_footer(
		    text=
		    'TickrBot by Group #24 of SJSU - Sp\'22.\nDo your own due diligence if you choose to follow any information given by TickrBot. Any action is of your own accord. Group #24 and its affiliates are not liable for any losses you may incur.'
		)

		await ctx.send(embed=embed)


# !chart <symbol> <duration> function
@bot.command(name='chart')
async def charting(ctx, symbol, duration):
	message = c.get_chart(symbol, duration)
	if message != True:
		await ctx.send(embed=message)
	else:
		filename = f'chart_{symbol}{duration}.png'
		with open(f'charts/chart_{symbol}{duration}.png', 'rb') as handle:
			image_send = discord.File(handle, filename=filename)
		await ctx.send(file=image_send)
		os.remove(f'charts/{filename}')


# !quote <symbol>
@bot.command(name='quote')
async def quoting(ctx, symbol):
	embed = q.get_quote(symbol)
	await ctx.send(embed=embed)


#================================== FUN STUFF =======================================================================
@bot.command('sheesh')
async def sheesh(ctx):
	desc = [
	    '(っ・ω・）っ≡≡≡≡≡≡.*･☆', '[^._.^]ﾉ', '( ˘•ω• ˘ )', '(´⊙ω⊙` )ᵒᵐᵍᵎᵎᵎ',
	    '┗(･ω･; )┛', '@(´ ･ｪ･ `)@'
	]
	desc = random.choice(desc)
	embed = discord.Embed(title="SHEEeeeeEEESH!!!!!",
	                      description=desc,
	                      color=discord.Color.random())
	await ctx.send(embed=embed)


@bot.command('hot')
async def hot(ctx):
	desc = ['heads', 'tails']
	desc = random.choice(desc)
	embed = discord.Embed(title=desc, color=discord.Color.greyple())
	await ctx.send(embed=embed)


bot.run(os.environ['h_token'])
